### PR TITLE
UCP/EP/FT: handle the case when no more lanes to fallback 

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What?
Handle the case when no more lanes to fallback, fixes:
- recursive request restart from zcopy completion
- am header leak
- ep_config ref counter fix for ifaces reactivation in lanes discard flow

## Why?
to fix various failures like hang, segfaults and assertions

### CODE REVIEW

File: src/ucp/am/eager_multi.c [hunk 1]
```diff
@@ -399,6 +403,15 @@ static ucs_status_t ucp_am_eager_multi_zcopy_psn_init(ucp_request_t *req)
         return status;
     }
 
+    /**
+     * At least 1 non-failed AM lane must be available.
+     */
+    if (ucp_ep_get_am_lane(req->send.ep) == UCP_NULL_LANE) {
```
Comment (yosefe) [inline src/ucp/am/eager_multi.c:409]: can we avoid extra branch in data path?

File: src/ucp/am/eager_multi.c [hunk 2]
```diff
@@ -380,7 +380,11 @@ static UCS_F_ALWAYS_INLINE ucs_status_t ucp_am_eager_multi_zcopy_psn_send_func(
 
 static void ucp_am_eager_multi_zcopy_psn_completion(uct_completion_t *self)
 {
-    if (ucs_likely(self->status == UCS_OK)) {
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t,
+                                          send.state.uct_comp);
+
+    if (ucs_likely(self->status == UCS_OK) ||
+        (req->send.ep->flags & UCP_EP_FLAG_FAILED)) {
```
Comment (yosefe) [inline src/ucp/am/eager_multi.c:387]: why needed, do we set flag FAILED in case of failover?

File: src/ucp/am/eager_multi.c [hunk 3]
```diff
@@ -430,7 +443,12 @@ static ucs_status_t ucp_am_eager_multi_zcopy_psn_reset(ucp_request_t *req)
     ucp_datatype_iter_rewind(&req->send.state.dt_iter, UCP_DT_MASK_CONTIG_IOV);
     /* Restart the request from the very first fragment */
     req->send.msg_proto.am.internal_flags &= ~UCP_REQUEST_AM_FLAG_HEADER_SENT;
-    return status;
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    return (ucp_ep_get_am_lane(req->send.ep) == UCP_NULL_LANE) ?
```
Comment (yosefe) [inline src/ucp/am/eager_multi.c:450]: what does it mean if ucp_ep_get_am_lane(req->send.ep) == UCP_NULL_LANE?

File: test/gtest/ucp/test_ucp_fault_tolerance.cc [hunk 4]
```diff
@@ -185,8 +185,18 @@ class test_ucp_fault_tolerance : public test_ucp_memheap {
...
                                   << ucs_status_string(status);
 
+        size_t num_lanes_to_fail = am_bw_lanes.size();
+        if (has_any_transport({"ud_v", "ud_x"}) &&
+            (failure_side == FAILURE_SIDE_INITIATOR)) {
+            /* TODO: remove this once UD ep purge assertions are fixed */
+            UCS_TEST_MESSAGE << "Keep 1 live lane for UD transports since "
+                             << "local error injection on all lanes leads to "
+                                "failed assertion in ud_ep_purge";
+            num_lanes_to_fail--;
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_fault_tolerance.cc:195]: Can `num_lanes_to_fail` be 0 or 1, before the subtraction?

File: test/gtest/ucp/test_ucp_fault_tolerance.cc [hunk 5]
```diff
@@ -200,14 +210,30 @@ class test_ucp_fault_tolerance : public test_ucp_memheap {
...
-                                      << ucs_status_string(status);
+            if (lane_idx < (am_bw_lanes.size() - 1)) {
+                EXPECT_EQ(UCS_OK, status) << op_str << " operation returned status: "
+                                        << ucs_status_string(status);
+                ASSERT_EQ(0, m_err_count) << "Error callback invoked " << m_err_count << " times";
+            } else {
+                short_progress_loop();
+                size_t min_expected_err_count = 1;
+                if ((failure_side == FAILURE_SIDE_TARGET) &&
+                    has_transport("dc_x")) {
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_fault_tolerance.cc:229]: Please document DC special case

File: src/ucp/core/ucp_ep.c [hunk 6]
```diff
@@ -1731,6 +1739,9 @@ ucp_ep_failover_reconfig(ucp_ep_h ucp_ep, ucp_lane_map_t failed_lanes,
 
     status = ucp_ep_reconfig_internal(ucp_ep, failed_lanes);
     if (status != UCS_OK) {
+        ucs_assertv(ucp_ep->cfg_index == old_cfg_index,
+                    "ep %p: cfg_index %u -> %u after reconfiguration error %s", ucp_ep, old_cfg_index,
```
Comment (gleon99) [inline src/ucp/core/ucp_ep.c:1743]: Minor: long line

File: src/ucp/am/eager.inl [hunk 7]
```diff
@@ -75,6 +75,7 @@ ucp_am_eager_zcopy_pack_user_header(ucp_request_t *req)
         return UCS_ERR_NO_MEMORY;
     }
 
+    ucp_trace_req(req, "allocating AM header descriptor 0x%p", reg_desc);
```
Comment (gleon99) [inline src/ucp/am/eager.inl:78]: Minor: `0x` is redundant.

File: test/gtest/ucp/test_ucp_fault_tolerance.cc [hunk 8]
```diff
@@ -30,7 +30,7 @@ class test_ucp_fault_tolerance : public test_ucp_memheap {
                                op_name(TEST_OP_GET | TEST_OP_FLUSH));
         add_variant_with_value(variants, UCP_FEATURE_AM,  TEST_OP_AM,
                                op_name(TEST_OP_AM));
-        add_variant_with_value(variants, UCP_FEATURE_AM | UCP_FEATURE_RMA,  TEST_OP_AM | TEST_OP_FLUSH,
+        add_variant_with_value(variants, UCP_FEATURE_AM,  TEST_OP_AM | TEST_OP_FLUSH,
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_fault_tolerance.cc:33]: Comment?

File: src/ucp/am/eager_multi.c [hunk 9]
```diff
@@ -430,7 +443,12 @@ static ucs_status_t ucp_am_eager_multi_zcopy_psn_reset(ucp_request_t *req)
...
+           UCS_ERR_UNREACHABLE : UCS_OK;
```
Comment (gleon99) [inline src/ucp/am/eager_multi.c:451]: The func returns OK if `ucp_am_proto_request_zcopy_reset` is OK, AND no AM lanes?
Needs at least commenting..

File: test/gtest/ucp/test_ucp_fault_tolerance.cc [hunk 10]
```diff
@@ -200,14 +210,30 @@ class test_ucp_fault_tolerance : public test_ucp_memheap {
...
+            if (lane_idx < (am_bw_lanes.size() - 1)) {
+                EXPECT_EQ(UCS_OK, status) << op_str << " operation returned status: "
+                                        << ucs_status_string(status);
+                ASSERT_EQ(0, m_err_count) << "Error callback invoked " << m_err_count << " times";
+            } else {
+                short_progress_loop();
+                size_t min_expected_err_count = 1;
+                if ((failure_side == FAILURE_SIDE_TARGET) &&
+                    has_transport("dc_x")) {
+                    min_expected_err_count = 0;
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_fault_tolerance.cc:230]: Can skip the test, the next assert is always true.

File: test/gtest/ucp/test_ucp_fault_tolerance.cc [hunk 11]
```diff
@@ -324,6 +350,7 @@ class test_ucp_fault_tolerance : public test_ucp_memheap {
     }
 
     ucs_status_t do_am_send_and_wait(ucp_ep_h ep, size_t size, bool flush_after) {
+
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_fault_tolerance.cc:353]: minor: remove

File: src/ucp/am/eager_multi.c (hunk 2 above)
Comment (evgeny-leksikov) [reply src/ucp/am/eager_multi.c:387]: Only when all lanes are failed

Comment (evgeny-leksikov) [reply src/ucp/am/eager_multi.c:387]: Needed to avoid leak of descriptor with copy of user AM header

File: test/gtest/ucp/test_ucp_fault_tolerance.cc (hunk 4 above)
Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_fault_tolerance.cc:195]: No, `shuffle_lanes` at very begin skips the test if num_lanes < 2

File: test/gtest/ucp/test_ucp_fault_tolerance.cc (hunk 8 above)
Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_fault_tolerance.cc:33]: `UCP_FEATURE_RMA` was added by mistaken earlier for AM test. Do you want to open separate PR for this trivial fix?

File: src/ucp/am/eager_multi.c (hunk 3 above)
Comment (evgeny-leksikov) [reply src/ucp/am/eager_multi.c]: It means that the are no live AM lanes after reconfiguration.
I converted such places to assertion since proto selection logic should not lead to this situation.

File: test/gtest/ucp/test_ucp_fault_tolerance.cc [hunk 12]
```diff
@@ -200,14 +210,31 @@ class test_ucp_fault_tolerance : public test_ucp_memheap {
...
+            if (lane_idx < (am_bw_lanes.size() - 1)) {
+                EXPECT_EQ(UCS_OK, status) << op_str << " operation returned status: "
+                                        << ucs_status_string(status);
+                ASSERT_EQ(0, m_err_count) << "Error callback invoked " << m_err_count << " times";
+            } else {
+                // The last lane is expected to fail
+                short_progress_loop();
+                if ((failure_side == FAILURE_SIDE_TARGET) &&
+                    has_transport("dc_x")) {
+                    // DC transport is not able to detect failure of remote DCI since DC is a connect2iface transport.
```
Comment (gleon99) [inline test/gtest/ucp/test_ucp_fault_tolerance.cc:230]: Empty block?

Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_fault_tolerance.cc:230]: why empty? the block describes test limitation in a corner case.

**Review by gleon99 — APPROVED** [review]

**Comment by gleon99** [comment]
@evgeny-leksikov please squash & merge
